service/sharddistributor/handler/executor_test.go

```diff
@@ -606,3 +610,241 @@ func newMigrationConfig(t *testing.T, configEntries []configEntry) *config.Migra
 	migrationConfig := config.NewMigrationConfig(dc)
 	return migrationConfig
 }
+
+func TestFilterNewlyAssignedShardIDs(t *testing.T) {
+	type testCase struct {
+		name     string
+		previous *store.HeartbeatState
+		assigned *store.AssignedState
+		expected []string
+	}
+	tests := []testCase{
+		{
+			name:     "nil previousHeartbeat returns all assigned",
+			previous: nil,
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{
+					"shard1": {},
+					"shard2": {},
+				},
+			},
+			expected: []string{"shard1", "shard2"},
+		},
+		{
+			name: "no new assigned shards",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+					"shard2": {},
+				},
+			},
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{
+					"shard1": {},
+					"shard2": {},
+				},
+			},
+			expected: []string{},
+		},
+		{
+			name: "some new assigned shards",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+				},
+			},
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{
+					"shard1": {},
+					"shard2": {},
+					"shard3": {},
+				},
+			},
+			expected: []string{"shard2", "shard3"},
+		},
+		{
+			name: "empty assigned returns empty",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+				},
+			},
+			assigned: &store.AssignedState{
+				AssignedShards: map[string]*types.ShardAssignment{},
+			},
+			expected: []string{},
+		},
+		{
+			name: "nil assignedState returns nil",
+			previous: &store.HeartbeatState{
+				ReportedShards: map[string]*types.ShardStatusReport{
+					"shard1": {},
+				},
+			},
+			assigned: nil,
+			expected: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result := filterNewlyAssignedShardIDs(tt.previous, tt.assigned)
+			require.ElementsMatch(t, tt.expected, result)
+		})
+	}
+}
+
+func TestEmitShardAssignmentMetrics(t *testing.T) {
```

Comment:
I think we should consider having some helpers to generate the `AssignedShards` and `ShardHandoverStats`. e.g: 
```go
func makeAssignedShards(shardIDs ...string) map[string]*types.ShardAssignment {
    m := make(map[string]*types.ShardAssignment, len(shardIDs))
    for _, id := range shardIDs {
        // Default to READY since that is the common case in your tests
        m[id] = &types.ShardAssignment{Status: types.AssignmentStatusREADY}
    }
    return m
}
```